### PR TITLE
Account import fixes

### PR DESCRIPTION
**Pull request by andrewkryshtal**

### CODE REVIEW

File: src/modals/import-account-modal/import-account-seed/import-account-seed.tsx [hunk 1]
```diff
@@ -87,6 +83,14 @@ export const ImportAccountSeed: FC<Props> = ({ onBackHandler }) => {
                 <FormTextInput name="derivationPath" />
               </>
             )}
+            <Divider size={formatSize(12)} />
+            <Label
+              label="Password"
+              isOptional
+              description={'Used for additional mnemonic derivation.\nThat is NOT a wallet password.'}
```
Comment (unixvb) [inline src/modals/import-account-modal/import-account-seed/import-account-seed.tsx:90]: `description="blabla"` stylegide

File: src/modals/import-account-modal/import-account-seed/import-account-seed.form.tsx [hunk 2]
```diff
@@ -22,5 +22,5 @@ export const importAccountSeedValidationSchema: SchemaOf<ImportAccountSeedValues
   seedPhrase: seedPhraseValidation,
   password: string(),
   derivationType: mixed<ImportAccountDerivationEnum>().oneOf(Object.values(ImportAccountDerivationEnum)).required(),
-  derivationPath: string().required()
+  derivationPath: string()
```
Comment (unixvb) [inline src/modals/import-account-modal/import-account-seed/import-account-seed.form.tsx:25]: need's validation for derivation path. 
should start with `m` if provided, valid syntax (take an example from extension)

File: src/shelter/use-shelter.hook.ts [hunk 3]
```diff
@@ -56,15 +57,32 @@ export const useShelter = () => {
...
-              title: 'Failed to import account.',
-              description: 'This may happen because provided Key is invalid.'
-            });
+          switchMap(({ privateKey, name }) => {
+            return getPublicKeyAndHash$(privateKey)
+              .pipe(
+                switchMap(([publicKey]) => {
+                  for (const account of accounts) {
+                    if (account.publicKey === publicKey) {
+                      showWarningToast({ description: 'Account already exist' });
```
Comment (unixvb) [inline src/shelter/use-shelter.hook.ts:67]: move "account already exist logic" into a component where `createImportedAccount` is called

File: src/shelter/use-shelter.hook.ts [hunk 4]
```diff
@@ -56,15 +57,32 @@ export const useShelter = () => {
...
+              )
+              .pipe(
+                catchError(() => {
+                  showErrorToast({
+                    title: 'Failed to import account.',
+                    description: 'This may happen because provided Key is invalid.'
+                  });
+
+                  return EMPTY;
+                })
```
Comment (unixvb) [inline src/shelter/use-shelter.hook.ts:84]: move this error handling into subsciption

File: src/modals/import-account-modal/import-account-seed/import-account-seed.tsx [hunk 5]
```diff
@@ -45,7 +46,9 @@ export const ImportAccountSeed: FC<Props> = ({ onBackHandler }) => {
     const seed = mnemonicToSeedSync(values.seedPhrase);
     const privateKey = seedToHDPrivateKey(
       seed,
-      values.derivationType === ImportAccountDerivationEnum.DEFAULT ? getDerivationPath(0) : values.derivationPath
+      values.derivationType === ImportAccountDerivationEnum.DEFAULT || !isString(values.derivationPath)
+        ? getDerivationPath(0)
+        : values.derivationPath
```
Comment (unixvb) [inline src/modals/import-account-modal/import-account-seed/import-account-seed.tsx:51]: does it's correct that you use `getDerivationPath(0)` when `!isString(values.derivationPath)`?  
As I remember it's different logic on extension.

File: src/shelter/use-shelter.hook.ts (hunk 3 above)
Comment (andrewkryshtal) [reply src/shelter/use-shelter.hook.ts:67]: i'm not agree. Here's one place with account checking, and we don't need to reuse shelter method to define `publicKey` and check if account exist. In your case - we should make some complicated steps on submit. Also,  `getPublicKeyAndHash` method is observable. What do you think about this?

**Review by keshan3262 — APPROVED** [review]

File: src/shelter/use-shelter.hook.ts [hunk 6]
```diff
@@ -56,15 +57,21 @@ export const useShelter = () => {
...
       createImportedAccount$
-        .pipe(switchMap(({ privateKey, name }) => Shelter.createImportedAccount$(privateKey, name)))
         .pipe(
-          catchError(() => {
-            showErrorToast({
-              title: 'Failed to import account.',
-              description: 'This may happen because provided Key is invalid.'
-            });
+          switchMap(({ privateKey, name }) => {
+            return getPublicKeyAndHash$(privateKey).pipe(
```
Comment (unixvb) [inline src/shelter/use-shelter.hook.ts:62]: short return

File: src/shelter/use-shelter.hook.ts [hunk 7]
```diff
@@ -73,6 +80,11 @@ export const useShelter = () => {
             dispatch(addHdAccountAction(publicData));
             showSuccessToast({ description: 'Account Imported!' });
             goBack();
+          } else {
+            showErrorToast({
+              title: 'Failed to import account.',
+              description: 'This may happen because provided Key is invalid.'
+            });
```
Comment (unixvb) [inline src/shelter/use-shelter.hook.ts:87]: add unit test for this case

File: src/utils/keys.utils.spec.ts [hunk 8]
```diff
@@ -1,20 +1,20 @@
...
 
-it('seedToHDPrivateKey should return private key, passing seed and derivation path', () => {
+it('seedToPrivateKey should return private key, passing seed and derivation path', () => {
   expect(
-    seedToHDPrivateKey(new Buffer(mockAccountCredentials.seedPhrase), mockAccountCredentials.derivationPath)
+    seedToPrivateKey(new Buffer(mockAccountCredentials.seedPhrase), mockAccountCredentials.derivationPath)
   ).toEqual(mockAccountCredentials.privateKey);
 });
 
 it('seedToHDPrivateKey should return error, passing seed without derivation path', () => {
```
Comment (unixvb) [inline src/utils/keys.utils.spec.ts:15]: change description

File: src/modals/import-account-modal/import-account-seed/import-account-seed.form.tsx [hunk 9]
```diff
@@ -8,7 +8,7 @@ export type ImportAccountSeedValues = {
   seedPhrase: string;
   password?: string;
   derivationType: ImportAccountDerivationEnum;
-  derivationPath: string;
+  derivationPath?: string;
```
Comment (unixvb) [inline src/modals/import-account-modal/import-account-seed/import-account-seed.form.tsx:11]: revert this. `derivationPath` could not be `undefined`

File: jest.setup.js [hunk 10]
```diff
@@ -12,8 +13,6 @@ if (typeof atob === 'undefined') {
 import './src/mocks/mem.mock';
 import './src/mocks/crypto.mock';
 
-import './src/mocks/taquito-signer.mock';
-
```
Comment (unixvb) [inline jest.setup.js:16]: revert this changes

File: src/shelter/use-shelter.hook.spec.ts [hunk 11]
```diff
@@ -108,15 +109,23 @@ describe('useShelter', () => {
...
     expect(mockUseDispatch).toBeCalledWith(setSelectedAccountAction(mockHdAccount.publicKeyHash));
     expect(mockUseDispatch).toBeCalledWith(addHdAccountAction(mockHdAccount));
     expect(mockShowSuccessToast).toBeCalledWith({ description: 'Account Imported!' });
     expect(mockGoBack).toBeCalled();
   });
+
+  it('should not create account with invalid private key', () => {
+    const { result } = renderHook(() => useShelter());
+    result.current.createImportedAccount({ privateKey: 'invalid privateKey', name: mockHdAccount.name });
+    expect(mockShelter.createImportedAccount$).not.toBeCalled();
```
Comment (unixvb) [inline src/shelter/use-shelter.hook.spec.ts:129]: expect(mockІhowErrorToast).toBeCalledWith(....);

File: src/modals/import-account-modal/import-account-seed/import-account-seed.form.tsx [hunk 12]
```diff
@@ -22,5 +22,26 @@ export const importAccountSeedValidationSchema: SchemaOf<ImportAccountSeedValues
...
+        return parts.every((p: string) => {
+          const pNum = +(p.includes("'") ? p.replace("'", '') : p);
+
+          return Number.isSafeInteger(pNum) && pNum >= 0;
+        });
+      }
+
+      return false;
+    })
+    .required()
```
Comment (unixvb) [inline src/modals/import-account-modal/import-account-seed/import-account-seed.form.tsx:46]: would this validation work with an empty string?

**Review by unixvb — APPROVED** [review]
